Refactoring _wrap_outputs to remove python dependence.

malvika2147 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22631 Refactoring _wrap_outputs to remove python dependence.**

Summary:
Created a python oblivious _wrap_outputs function by refactoring the original one. The goal is to refactor all of python_function.cpp to remove dependence on python, so that we can create a C++ wrapper for it.

Test Plan: test suite

Differential Revision: [D16185040](https://our.internmc.facebook.com/intern/diff/D16185040)

Commit:
Refactoring _wrap_outputs to remove python dependence.

Summary:
Created a python oblivious _wrap_outputs function.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,80 @@
+
+TORCH_API std::vector<Variable> _wrap_outputs(std::unordered_set<Variable*> &inputs,
+  std::vector<Variable*> &raw_outputs, std::unordered_set<Variable*> &non_differentiable,
+  std::vector<Variable*> &dirty_inputs, std::shared_ptr<Function> cdata) {
```
We don't really like non-const references, so if you don't need to actually write to these, make them const. (I feel like inputs, for example, should be const, at the very least?)

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,80 @@
+  std::vector<Variable*> &raw_outputs, std::unordered_set<Variable*> &non_differentiable,
+  std::vector<Variable*> &dirty_inputs, std::shared_ptr<Function> cdata) {
+  // Sets the grad_fn and output_nr of an output Variable.
+  auto set_history = [&](Variable& var, uint32_t output_nr, bool is_input, bool is_modified,
```
Similar here, this can probably be a `const Variable&`

ezyang replied:
OK, I'm wrong, this can't be const

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -480,10 +431,10 @@ static void _save_variables(THPFunction* self)
 _parse_non_differentiable(THPFunction *self)
 {
-  std::unordered_set<PyObject*> set;
+  std::unordered_set<Variable*> set;
```
This is very unlikely to be the right thing. Variable itself is a pointer type; if you need a pointer to hash on, probably the right thing is to call `unsafeGetTensorImpl()` which will give you a `TensorImpl*` which is appropriate to hash on.

malvika2147 replied:
It doesn't have to be a pointer, I can have a set of Variable instead.

ezyang replied:
What you've done here is hash on the location of the `Variable` (i.e., `TensorImpl**`), this accidentally works because you always take an address from a PyObject which then uniquely identifies the memory location of `Variable`

ezyang replied:
Sure, a set of Variable is good to, assuming they are hashable (I'm not sure they are. But it would be a reasonable thing for them to be hashable.)

ezyang commented:
> Created a python oblivious _wrap_outputs function.

For the benefit of people who are not me, a more detailed PR description will be helpful :)

ezyang commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -362,89 +363,39 @@ static void _wrap_outputs(THPFunction *self,
   }
 
-  std::unordered_set<PyObject*> inputs;
+  auto as_variable = [&](PyObject* obj, TypeError err) -> Variable* {
```
Similarly here, just return a Variable. (mumble mumble extra refcount bump mumble)

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,80 @@
+    }
+  };
+
+  std::vector<torch::autograd::Variable> outputs;
```
Micro optimization tip: reserve the correct number of output elements, so we don't have to reallocate the buffer

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,80 @@
+    }
+    set_history(var, i, is_input, is_modified, is_differentiable);
+
+    if (cdata) {
```
Are you sure?! I would have expected to add the variable to outputs always

ezyang reviewed (dismissed):
Still some work to do. I'll need to do a closer review after improvements.

Commit:
Update on "[WIP] Refactoring _wrap_outputs to remove python dependence."

Refactoring _wrap_outputs to remove python dependence.

Summary:
Created a python oblivious _wrap_outputs function.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22631 gh/mal2147/12/head

Commit:
Update on "[WIP] Refactoring _wrap_outputs to remove python dependence."

Refactoring _wrap_outputs to remove python dependence.

Summary:
Created a python oblivious _wrap_outputs function.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22631 gh/mal2147/12/head

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,84 @@
+
+namespace torch { namespace autograd {
+
+std::vector<Variable> _wrap_outputs(const std::unordered_set<at::TensorImpl*> &inputs,
```
A useful C++ tip to know for the future: `unordered_map` and `unordered_set` are actually really crappy hash map implementations and perform pretty badly, and there is usually something faster in reach. We have one in `c10/util/flat_hash_map.h`. I don't know if it matters here, since this code isn't that fast anyway, but if you ever are in a situation where you need a hash map, there you go.

ezyang replied:
(And the old code used `unordered_set` anyway, so, it's no worse!)

ezyang commented on torch/csrc/autograd/custom_function.h:
```diff
@@ -0,0 +1,15 @@
+
+namespace torch { namespace autograd {
+
+TORCH_API std::vector<Variable> _wrap_outputs(
```
Do you wanna write some docs for this function? :) (Doesn't have to be this PR)

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,84 @@
+  const std::unordered_set<at::TensorImpl*> &non_differentiable,
+  const std::unordered_set<at::TensorImpl*> &dirty_inputs,
+  const std::vector<Variable> &raw_outputs,
+  std::shared_ptr<Function> cdata) {
```
C++ nit: you can avoid the refcount bump by passing this shared pointer by const reference. I also don't think it matters here, but it's a pretty common idiom in C++ and worth knowing about (since lots of refcount bumps add up!)

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,84 @@
+std::vector<Variable> _wrap_outputs(const std::unordered_set<at::TensorImpl*> &inputs,
+  const std::unordered_set<at::TensorImpl*> &non_differentiable,
+  const std::unordered_set<at::TensorImpl*> &dirty_inputs,
+  const std::vector<Variable> &raw_outputs,
```
C++ in our codebase nit: it's more idiomatic to take an `ArrayRef<Variable> raw_outputs` here. ArrayRef is our general purpose "reference to vector-like struct"; it can take out a reference to vectors, but other formats too.

ezyang commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -0,0 +1,84 @@
+  for (auto i = 0; i < num_outputs; ++i) {
+    auto out_tensor_impl = raw_outputs[i].unsafeGetTensorImpl();
+    bool is_input = inputs.count(out_tensor_impl) > 0;
+    bool is_modified = std::find(dirty_inputs.begin(), dirty_inputs.end(), out_tensor_impl) != dirty_inputs.end();
```
Err, you shouldn't use `std::find` anymore, now that you made `dirty_inputs` a set, right? You can just use `count` directly now.

ezyang reviewed (approved):
There's some nits, but the only that really stands out is the use of `std::find` on an unordered map (please don't do that!) Everything else is just nits.

Commit:
Update on "Refactoring _wrap_outputs to remove python dependence."

Refactoring _wrap_outputs to remove python dependence.

Summary:
Created a python oblivious _wrap_outputs function.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22631 gh/mal2147/12/head

Commit:
Update on "Refactoring _wrap_outputs to remove python dependence."

Refactoring _wrap_outputs to remove python dependence.

Summary:
Created a python oblivious _wrap_outputs function.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

gh-metadata: pytorch pytorch 22631 gh/mal2147/12/head
